internal/engine/dockerprune/docker_pruner.go

```diff
@@ -63,58 +63,59 @@ func (dp *DockerPruner) SetUp(ctx context.Context, _ store.RStore) error {
 	return nil
 }
 
-func (dp *DockerPruner) OnChange(ctx context.Context, st store.RStore, _ store.ChangeSummary) error {
-	if dp.disabledForTesting || dp.disabledOnSetup {
+// OnChange determines if any Tilt-built Docker images should be pruned based on settings and invokes the pruning
+// process if necessary.
+//
+// Care should be taken when modifying this method to not introduce expensive operations unless necessary, as this
+// is invoked for EVERY store action change batch. Because of this, the store (un)locking is done somewhat manually,
+// so care must be taken to avoid locking issues.
+func (dp *DockerPruner) OnChange(ctx context.Context, st store.RStore, summary store.ChangeSummary) error {
+	if dp.disabledForTesting || dp.disabledOnSetup || summary.IsLogOnly() {
 		return nil
 	}
 
 	state := st.RLockState()
 	settings := state.DockerPruneSettings
-	buildInProg := len(state.CurrentlyBuilding) > 0
-	curBuildCount := state.CompletedBuildCount
-	hasDockerBuild := state.HasDockerBuild()
-	nextToBuild := buildcontrol.NextManifestNameToBuild(state)
-	imgSelectors := model.LocalRefSelectorsForManifests(state.Manifests())
-	st.RUnlockState()
-
-	if !settings.Enabled {
-		return nil
-	}
-
-	// If user doesn't have at least one Docker build, they probably don't care about pruning
-	if !hasDockerBuild {
-		return nil
-	}
-
-	// Don't prune while we're building or about to build something, in case of weird side-effects.
-	if buildInProg || nextToBuild != "" {
+	// Exit early if possible if any of the following is true:
+	// 	* Pruning is disabled entirely
+	// 	* Engine is currently building something
+	// 	* There are NO `docker_build`s in the Tiltfile
+	// 	* Something is queued for building
+	if !settings.Enabled || len(state.CurrentlyBuilding) > 0 || !state.HasDockerBuild() || buildcontrol.NextManifestNameToBuild(state) != "" {
+		st.RUnlockState()
```

Comment:
can these all be moved into a `defer` after the `RLockState`?

Reply:
The issue would be holding the lock during the _actual_ call to `Prune`.

That said, since subscribers already run non-blocking in their own goroutines, we could fire off the actual prune in a goroutine [would need to add some slight extra synchronization then to make sure we don't run it multiple times simultaneously].

Or could do something kinda hacky with a `sync.Once(st.RUnlockState())` - call it via `defer` or early in the path that actually prunes.

Thoughts?

Reply:
I'd lean towards that all being more complexity than it's worth. I missed the lock on Prune issue.